Adding IResult docs to MVC (#27157)
dotnet/AspNetCore.Docs: aspnetcore/web-api/action-return-types.md

In the preceding action:

* A 400 status code (<xref:Microsoft.AspNetCore.Mvc.ControllerBase.BadRequest%2A>) is returned by the ASP.NET Core runtime when:
  * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute has been applied and model validation fails.
  * The product description contains "XYZ Widget".
* A 201 status code is generated by the <xref:Microsoft.AspNetCore.Mvc.ControllerBase.CreatedAtAction%2A> method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.

## HttpResults type

TODO

<!-- Mention no content-negotiation -->

### IResult

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/IResultProductsController.cs" id="snippet_GetByIdIResult" highlight="7":::

In the preceding action:

* A 404 status code is returned when the product doesn't exist in the database.

> * A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [Results.Ok\<T>()](<xref:Microsoft.AspNetCore.Http.Results.Ok>).

Reviewer: @guardrex  your tool has `<xref:Microsoft.AspNetCore.Http.Results.Ok(Object)>`

Reviewer 2: It doesn't work very well any longer. I think that there were changes on the API end that I never reacted to. I don't currently have plans to work on it further. Is the one that Kirk made still running? I had the impression that his version was providing better (*correct!*) links.

Reviewer: I think it's https://github.com/serpent5/Serpent5.Xrefs 
I had trouble deploying it. @guardrex  maybe you could help me get it deployed.
cc @serpent5

Reviewer 2: ![mccoy](https://user-images.githubusercontent.com/1622880/193427376-0f5dc3a3-7dae-4cfb-8d74-a1b2d699a927.png)

### *Damn it Rick! I'm a .NET guy, not a Node JSX dev!* 🤣

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&mdash; Dr. Leonard H. "Bones" McCoy ([DeForest Kelley](https://en.wikipedia.org/wiki/DeForest_Kelley)) [Star Trek](http://www.startrek.com/), ©1966-69 [Paramount Pictures](http://www.paramount.com)

Reviewer 3: Yeah, that's the repo. It was deployed as part of my MSFT subscription, but that was deleted. 

@Rick-Anderson I'll email you tomorrow with some instructions on how to deploy it.

Reviewer 2: Thx @serpent5 ... That *RexHackCrap*&trade; that I built has broken down. If Rick gets your app running, I'll start using that. I don't have the time or interest to fix the one that I built.

Reviewer: > Yeah, that's the repo. It was deployed as part of my MSFT subscription, but that was deleted.
> 
Might be easier to get you MVP status again :)

> * A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [Results.Ok\<T>()](<xref:Microsoft.AspNetCore.Http.Results.Ok>).

Reviewer: ```suggestion
* A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [Results.Ok\<T>()](/dotnet/api/microsoft.aspnetcore.http.results.ok).
```

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/IResultProductsController.cs" id="snippet_CreateAsyncIResult" highlight="9,16":::

In the preceding action:

* A 400 status code is returned by the ASP.NET Core runtime when:
  * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute has been applied and model validation fails.
  * The product description contains "XYZ Widget".

> * A 201 status code is generated by the [Results.Create()](<xref:Microsoft.AspNetCore.Http.Results.Created>) method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.

Reviewer: ```suggestion
* A 201 status code is generated by the [`Results.Create`](/dotnet/api/microsoft.aspnetcore.http.results.created) method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.
```

### Result\<T> type

<!-- Add doc about the OpenAPI (Produces) metadata -->

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ResultsOfTProductsController.cs" id="snippet_GetByIdResultsOfT" highlight="5":::

In the preceding action:

* A 404 status code is returned when the product doesn't exist in the database.
* A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [TypedResults.Ok\<T>()](<xref:Microsoft.AspNetCore.Http.TypedResults.Ok>).

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ResultsOfTProductsController.cs" id="snippet_CreateAsyncResultsOfT" highlight="6,13":::

In the preceding action:

* A 400 status code is returned by the ASP.NET Core runtime when:
  * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute has been applied and model validation fails.
  * The product description contains "XYZ Widget".

> * A 201 status code is generated by the [TypedResults.Create()](<xref:Microsoft.AspNetCore.Http.TypedResults.Created>) method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.

Reviewer: ```suggestion
* A 201 status code is generated by the [`TypedResults.Create`](/dotnet/api/microsoft.aspnetcore.http.typedresults.created) method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.
```

## HttpResults type

In addition to the MVC-specific built-in result types (<xref:Microsoft.AspNetCore.Mvc.IActionResult> and [ActionResult\<T>](xref:Microsoft.AspNetCore.Mvc.ActionResult%601)), ASP.NET Core includes the [HttpResult](xref:Microsoft.AspNetCore.Http.HttpResults) type that can be used in both [Minimal APIs](https://learn.microsoft.com/aspnet/core/fundamentals/minimal-apis) and Web API.

Differently than the MVC-specific result types, the `HttpResults` are results implementation that will be processed by a call to [IResult.ExecuteAsync](xref:Microsoft.AspNetCore.Http.IResult.ExecuteAsync%2A) and *will not* leverage the configured [Formatters](https://learn.microsoft.com/aspnet/core/web-api/advanced/formatting#format-specific-action-results), that means, some features like `Content negotiation` are not available and the produced `Content-Type` will be decided by the `HttpResults` implementation.

The `HttpResults` can be useful when sharing code between both frameworks (Minimal APIs and Web API). 

### IResult type

> The `HttpResults` are a implementation of the <xref:Microsoft.AspNetCore.Http.IResult> interface, that defines a contract that represents the result of an HTTP endpoint, and the static [Results](<xref:Microsoft.AspNetCore.Http.Results>) class is used to create varying `IResult` objects that represent different types of responses.

Author: You change make sense, I was calling "HttpResults" all types in the Microsoft.AspNetCore.Http.HttpResults namespace

> This table shows the common result helpers available: [Built-in results](/aspnet/core/fundamentals/minimal-apis#built-in-results-1)

Reviewer: > Where can I find all possible `query params`? Eg. `preserve-view=true`

That's the main one I know about. If you don't add  `preserve-view=true`, the build report gives you a warning to add it or remove the version query string.

In the preceding action:

* A 400 status code is returned by the ASP.NET Core runtime when:
  * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute has been applied and model validation fails.
  * The product description contains "XYZ Widget".
* A 201 status code is generated by the [`Results.Create`](/dotnet/api/microsoft.aspnetcore.http.results.created) method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.

### Result\<T> type

ASP.NET Core also includes the static [TypedResults](<xref:Microsoft.AspNetCore.Http.TypedResults>) class that returns the concrete `IResult` implementation, that allows using it as return type. The usage of the concrete `IResult` implementation offers the following benefit over the [IResult type](#iresult-type):

* All the [`[ProducesResponseType]`](xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute) attribute's can be excluded, since the `HttpResult` implementation will contribute automatically to the endpoint metadata.

> When multiple `IResult` return types are needed, while still possible use `IResult` interface as return type is better use the [`Result<TResult1, TResultN>`](/dotnet/api/microsoft.aspnetcore.http.httpresults.results-2) generic union types that will keep the automatic endpoint metadata benefit.

Reviewer: ```suggestion
When multiple `IResult` return types are needed, returning [`Result<TResult1, TResultN>`](/dotnet/api/microsoft.aspnetcore.http.httpresults.results-2) is preferred over returning `IResult`. Returning `Result<TResult1, TResultN>` is preferred because generic union types automatically retain the endpoint metadata.
```

> The `Results<TResult1, TResultN>` union types implement implicit cast operators so that the compiler can automatically convert the types specified in the generic arguments to an instance of the union type. This has the added benefit of providing compile-time checking that a route handler actually only returns the results that it declares it does. Attempting to return a type that isn’t declared as one of the generic arguments to `Results<>` will result in a compilation error.

Reviewer: ```suggestion
The `Results<TResult1, TResultN>` union types implement implicit cast operators so that the compiler can automatically convert the types specified in the generic arguments to an instance of the union type. This has the added benefit of providing compile-time checking that a route handler actually only returns the results that it declares it does. Attempting to return a type that isn’t declared as one of the generic arguments to `Results<>` results in a compilation error.
```

Consider the following code:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ResultsOfTProductsController.cs" id="snippet_GetByIdResultsOfT" highlight="5":::

In the preceding action:

* A 404 status code is returned when the product doesn't exist in the database.

> * A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [TypedResults.Ok\<T>()](<xref:Microsoft.AspNetCore.Http.TypedResults.Ok>).

Reviewer: ```suggestion
* A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [TypedResults.Ok\<T>](<xref:Microsoft.AspNetCore.Http.TypedResults.Ok>).
```

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ResultsOfTProductsController.cs" id="snippet_CreateAsyncResultsOfT" highlight="6,13":::

In the preceding action:

> * A 400 status code is returned by the ASP.NET Core runtime when:

Reviewer: ```suggestion
* A 400 status code is returned when:
```

>   * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute has been applied and model validation fails.

Reviewer: ```suggestion
  * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute was applied and model validation fails.
```

### Return IEnumerable\<T> or IAsyncEnumerable\<T>

ASP.NET Core buffers the result of actions that return <xref:System.Collections.Generic.IEnumerable%601> before writing them to the response. Consider declaring the action signature's return type as <xref:System.Collections.Generic.IAsyncEnumerable%601> to guarantee asynchronous iteration. Ultimately, the iteration mode is based on the underlying concrete type being returned. MVC automatically buffers any concrete type that implements `IAsyncEnumerable<T>`.

Consider the following action, which returns sale-priced product records as `IEnumerable<Product>`:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ProductsController.cs" id="snippet_GetOnSaleProducts":::

The `IAsyncEnumerable<Product>` equivalent of the preceding action is:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ProductsController.cs" id="snippet_GetOnSaleProductsAsync":::

> 

Reviewer: This section doesn't provide any guidance on when to use async. It asks the question and doesn't answer it.
> ASP.NET Core buffers the result of actions that return `IEnumerable<T>`
> MVC automatically buffers any concrete type that implements `IAsyncEnumerable<T>.`

So they're both buffered.
Do we want to say anything about buffering vs streaming?  In general, streaming is preferred.
There's also `IEnumerable<Task<T>>`

Author: Actually, it does not buffer anymore, since .NET 6 (https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/6.0/iasyncenumerable-not-buffered-by-mvc). Let me update the text.

* A 200 status code is returned with the `Product` object when the product does exist. The <xref:Microsoft.AspNetCore.Mvc.ControllerBase.Ok%2A> convenience method is invoked as shorthand for `return new OkObjectResult(product);`.

### Asynchronous action

Consider the following asynchronous action in which there are two possible return types:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ActionResultProductsController.cs" id="snippet_CreateAsyncIActionResult" highlight="9,15":::

In the preceding action:

* A 400 status code is returned when the product description contains "XYZ Widget". The <xref:Microsoft.AspNetCore.Mvc.ControllerBase.BadRequest%2A> convenience method is invoked as shorthand for `return new BadRequestResult();`.
* A 201 status code is generated by the <xref:Microsoft.AspNetCore.Mvc.ControllerBase.CreatedAtAction%2A> convenience method when a product is created. The following code is an alternative to calling `CreatedAtAction`:

  ```csharp

>   return new CreatedAtActionResult(nameof(GetById), "Products", new { id = product.Id }, product);

Reviewer: ```suggestion
  return new CreatedAtActionResult(nameof(GetById), "Products",
                                                                    new { id = product.Id }, product);
```
Get rid of horizontal scroll bar
<img width="514" alt="image" src="https://user-images.githubusercontent.com/3605364/193922980-a43446fe-940c-4f52-9385-e8b6bc913350.png">

Different than the MVC-specific result types, the `HttpResults`:

* Are a results implementation that is processed by a call to [IResult.ExecuteAsync](xref:Microsoft.AspNetCore.Http.IResult.ExecuteAsync%2A).
* Does ***not*** leverage the configured [Formatters](/aspnet/core/web-api/advanced/formatting#format-specific-action-results). Not leveraging the configured formatters means:

  * Some features like `Content negotiation` aren't available.
  * The produced `Content-Type` is decided by the `HttpResults` implementation.

The `HttpResults` can be useful when sharing code between Minimal APIs and Web API. 

### IResult type

The <xref:Microsoft.AspNetCore.Http.HttpResults> namespace contains classes that implement the <xref:Microsoft.AspNetCore.Http.IResult> interface. The `IResult` interface defines a contract that represents the result of an HTTP endpoint. The static [Results](<xref:Microsoft.AspNetCore.Http.Results>) class is used to create varying `IResult` objects that represent different types of responses.

> The [Built-in results](/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0&preserve-view=true#bimr7) table shows the common result helpers.

Author: Is this link expected to work already? Still not pointing to the right header.

Reviewer: > Is this link expected to work already? Still not pointing to the right header.

No, you'd have to `git pull origin main` to bring it in, but when the PR merges, the link will work.

* [Specific type](#specific-type)
* [IActionResult](#iactionresult-type)
* [ActionResult\<T>](#actionresultt-type)
* [HttpResults](#httpresults-type)

This article explains when it's most appropriate to use each return type.

## Specific type

The most basic action returns a primitive or complex data type, for example, `string` or a custom object. Consider the following action, which returns a collection of custom `Product` objects:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ProductsController.cs" id="snippet_Get":::

Without known conditions to safeguard against, returning a specific type could suffice. The preceding action accepts no parameters, so parameter constraints validation isn't needed.

When multiple return types are possible, it's common to mix an <xref:Microsoft.AspNetCore.Mvc.ActionResult> return type with the primitive or complex return type. Either [IActionResult](#iactionresult-type) or [ActionResult\<T>](#actionresultt-type) are necessary to accommodate this type of action. Several samples of multiple return types are provided in this article.

> ### Return IEnumerable\<T> or IAsyncEnumerable\<T>

Reviewer: Should we S&M this PR and I'll open an issue on this doc?
`### Return IEnumerable\<T> or IAsyncEnumerable\<T>` never answers the question or provides guidance.

Author: We do have in the best practices docs https://learn.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-6.0#return-ienumerablet-or-iasyncenumerablet but sure let's open an issue so we can improve it.

Reviewer: @brunolins16  see #27210

> ASP.NET Core buffers the result of actions that return <xref:System.Collections.Generic.IEnumerable%601> before writing them to the response. Consider declaring the action signature's return type as <xref:System.Collections.Generic.IAsyncEnumerable%601> to guarantee asynchronous iteration. Ultimately, the iteration mode is based on the underlying concrete type being returned and the selected formatter affects how the result is processed:

Reviewer: > Ultimately, the iteration mode is based on the underlying concrete type being returned and the selected formatter affects how the result is processed:

how that impact sync vs async is not clear.
